File: frontend/src/types/sales.ts
export type SalesOrderStatus =
  | "DRAFT"
  | "CONFIRMED"
  | "PARTIALLY_DELIVERED"
  | "DELIVERED"
  | "CANCELLED";

export interface SalesOrderLine {
  id: number;
  product_id: number;
  ordered_quantity: number;
  delivered_quantity: number;
  sales_price: number;
  total: number;
  remaining_to_deliver: number;
}

/** SalesOrderOut from GET /sales-orders. */
export interface SalesOrder {
  id: number;
  customer_name: string;
  customer_address: string | null;
  salesperson: string | null;
  creation_date: string;
  status: SalesOrderStatus;
  total_amount: number;
  lines: SalesOrderLine[];
}

export interface SalesOrderLineCreate {
  product_id: number;
  ordered_quantity: number;
  sales_price?: number | null;
}

export interface SalesOrderCreate {
  customer_name: string;
  customer_address?: string | null;
  salesperson?: string | null;
  lines: SalesOrderLineCreate[];
}

export interface DeliveryLineInput {
  line_id: number;
  quantity: number;
}

/** Body for POST /sales-orders/{id}/deliver. Omit `lines` to deliver all remaining. */
export interface DeliveryRequest {
  lines?: DeliveryLineInput[];
}

/** Response of POST /sales-orders/{id}/confirm. */
export interface ConfirmationResult {
  sales_order: SalesOrder;
  generated_purchase_order_ids: number[];
  generated_manufacturing_order_ids: number[];
  messages: string[];
}

File: backend/schemas/sales.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from utils.enums import SalesOrderStatus


class SalesOrderLineCreate(BaseModel):
    product_id: int
    ordered_quantity: float = Field(gt=0)
    # If omitted, the service falls back to the product's sales_price.
    sales_price: float | None = Field(default=None, ge=0)


class SalesOrderCreate(BaseModel):
    customer_name: str = Field(min_length=1, max_length=200)
    customer_email: str | None = Field(default=None, max_length=200)
    customer_address: str | None = None
    salesperson: str | None = None
    lines: list[SalesOrderLineCreate] = Field(min_length=1)


class DeliveryLine(BaseModel):
    line_id: int
    quantity: float = Field(gt=0)


class DeliveryRequest(BaseModel):
    """Quantities to deliver per line. If omitted entirely, the service
    delivers all remaining quantity for every line."""
    lines: list[DeliveryLine] | None = None


class SalesOrderLineOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    product_id: int
    ordered_quantity: float
    delivered_quantity: float
    reserved_quantity: float
    sales_price: float
    total: float
    remaining_to_deliver: float


class SalesOrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    customer_name: str
    customer_email: str | None
    customer_address: str | None
    salesperson: str | None
    creation_date: datetime
    status: SalesOrderStatus
    total_amount: float
    lines: list[SalesOrderLineOut]


class ConfirmationResult(BaseModel):
    """Returned by confirm() so the frontend knows what automation fired."""
    sales_order: SalesOrderOut
    generated_purchase_order_ids: list[int] = []
    generated_manufacturing_order_ids: list[int] = []
    messages: list[str] = []
